File: accelerator/cre-custom-data-feed-go/workflow/workflow_test.go
package main

import (
	"context"
	_ "embed"
	"encoding/json"
	"math/big"
	"strings"
	"testing"
	"time"

	"github.com/ethereum/go-ethereum/common"
	"github.com/smartcontractkit/cre-sdk-go/capabilities/blockchain/evm"
	evmmock "github.com/smartcontractkit/cre-sdk-go/capabilities/blockchain/evm/mock"
	"github.com/smartcontractkit/cre-sdk-go/capabilities/networking/http"
	httpmock "github.com/smartcontractkit/cre-sdk-go/capabilities/networking/http/mock"
	"github.com/smartcontractkit/cre-sdk-go/capabilities/scheduler/cron"
	"github.com/smartcontractkit/cre-sdk-go/cre/testutils"
	"github.com/stretchr/testify/require"
	"google.golang.org/protobuf/types/known/timestamppb"

	"cre-v1-project/contracts/evm/src/generated/balance_reader"
	"cre-v1-project/contracts/evm/src/generated/ierc20"
	"cre-v1-project/contracts/evm/src/generated/message_emitter"
)

var anyExecutionTime = time.Unix(1752514917, 0)

func TestInitWorkflow(t *testing.T) {
	config := makeTestConfig(t)
	runtime := testutils.NewRuntime(t, map[string]string{})

	workflow, err := InitWorkflow(config, runtime.Logger(), nil)
	require.NoError(t, err)

	require.Len(t, workflow, 3) // cron, log, and http triggers
	require.Equal(t, cron.Trigger(&cron.Config{}).CapabilityID(), workflow[0].CapabilityID())
}

func TestOnCronTrigger(t *testing.T) {
	config := makeTestConfig(t)
	runtime := testutils.NewRuntime(t, map[string]string{
		"/SECRET_ADDRESS": "0x1234567890123456789012345678901234567890",
	})

	// Mock HTTP client for POR data
	httpMock, err := httpmock.NewClientCapability(t)
	require.NoError(t, err)
	httpMock.SendRequest = func(ctx context.Context, input *http.Request) (*http.Response, error) {
		// Return mock POR response
		porResponse := `{
			"accountName": "TrueUSD",
			"totalTrust": 1000000.0,
			"totalToken": 1000000.0,
			"ripcord": false,
			"updatedAt": "2023-01-01T00:00:00Z"
		}`
		return &http.Response{Body: []byte(porResponse)}, nil
	}

	// Mock EVM client
	chainSelector, err := config.EVMs[0].GetChainSelector()
	require.NoError(t, err)
	evmMock, err := evmmock.NewClientCapability(chainSelector, t)
	require.NoError(t, err)

	// Set up contract mocks using generated mock contracts
	evmCfg := config.EVMs[0]

	// Mock BalanceReader for fetchNativeTokenBalance
	balanceReaderMock := balance_reader.NewBalanceReaderMock(
		common.HexToAddress(evmCfg.BalanceReaderAddress),
		evmMock,
	)
	balanceReaderMock.GetNativeBalances = func(input balance_reader.GetNativeBalancesInput) ([]*big.Int, error) {
		// Return mock balance for each address (same number as input addresses)
		balances := make([]*big.Int, len(input.Addresses))
		for i := range input.Addresses {
			balances[i] = big.NewInt(500000000000000000) // 0.5 ETH in wei
		}
		return balances, nil
	}

	// Mock IERC20 for getTotalSupply
	ierc20Mock := ierc20.NewIERC20Mock(
		common.HexToAddress(evmCfg.TokenAddress),
		evmMock,
	)
	ierc20Mock.TotalSupply = func() (*big.Int, error) {
		return big.NewInt(1000000000000000000), nil // 1 token with 18 decimals
	}

	// Note: ReserveManager WriteReportFromUpdateReserves is not a read method,
	// so it's handled by the EVM mock transaction system directly
	evmMock.WriteReport = func(ctx context.Context, input *evm.WriteReportRequest) (*evm.WriteReportReply, error) {
		return &evm.WriteReportReply{
			TxHash: common.HexToHash("0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef").Bytes(),
		}, nil
	}

	result, err := onPORCronTrigger(config, runtime, &cron.Payload{
		ScheduledExecutionTime: timestamppb.New(anyExecutionTime),
	})

	require.NoError(t, err)
	require.NotNil(t, result)

	// Check that the result contains the expected reserve value
	require.Equal(t, "1000000", result) // Should match the totalToken from mock response

	// Verify expected log messages
	logs := runtime.GetLogs()
	assertLogContains(t, logs, `msg="fetching por"`)
	assertLogContains(t, logs, `msg=ReserveInfo`)
	assertLogContains(t, logs, `msg=TotalSupply`)
	assertLogContains(t, logs, `msg=TotalReserveScaled`)
	assertLogContains(t, logs, `msg="Native token balance"`)
	assertLogContains(t, logs, `msg="Secret address balance"`)
}

func TestOnLogTrigger(t *testing.T) {
	config := makeTestConfig(t)
	runtime := testutils.NewRuntime(t, map[string]string{})

	// Mock EVM client
	chainSelector, err := config.EVMs[0].GetChainSelector()
	require.NoError(t, err)
	evmMock, err := evmmock.NewClientCapability(chainSelector, t)
	require.NoError(t, err)

	// Mock MessageEmitter for log trigger
	evmCfg := config.EVMs[0]
	messageEmitterMock := message_emitter.NewMessageEmitterMock(
		common.HexToAddress(evmCfg.MessageEmitterAddress),
		evmMock,
	)
	messageEmitterMock.GetLastMessage = func(input message_emitter.GetLastMessageInput) (string, error) {
		return "Test message from contract", nil
	}

	// Create a mock log payload
	mockLog := &evm.Log{
		Topics: [][]byte{
			common.HexToHash("0x1234567890123456789012345678901234567890123456789012345678901234").Bytes(), // event signature
			common.HexToHash("0x000000000000000000000000abcdefabcdefabcdefabcdefabcdefabcdefabcd").Bytes(), // emitter address (padded)
			common.HexToHash("0x5678567856785678567856785678567856785678567856785678567856785678").Bytes(), // additional topic
		},
		Data: []byte{},
	}

	result, err := onLogTrigger(config, runtime, mockLog)
	require.NoError(t, err)
	require.Equal(t, "Test message from contract", result)

	// Verify expected log messages
	logs := runtime.GetLogs()
	assertLogContains(t, logs, `msg="Message retrieved from the contract"`)
}

func TestOnHTTPTrigger(t *testing.T) {
	config := makeTestConfig(t)
	runtime := testutils.NewRuntime(t, map[string]string{
		"/SECRET_ADDRESS": "0x9876543210987654321098765432109876543210",
	})

	// Mock HTTP client for POR data
	httpMock, err := httpmock.NewClientCapability(t)
	require.NoError(t, err)
	httpMock.SendRequest = func(ctx context.Context, input *http.Request) (*http.Response, error) {
		porResponse := `{
			"accountName": "TrueUSD",
			"totalTrust": 2000000.0,
			"totalToken": 2000000.0,
			"ripcord": false,
			"updatedAt": "2023-01-01T00:00:00Z"
		}`
		return &http.Response{Body: []byte(porResponse)}, nil
	}

	// Mock EVM client (same pattern as cron test)
	chainSelector, err := config.EVMs[0].GetChainSelector()
	require.NoError(t, err)
	evmMock, err := evmmock.NewClientCapability(chainSelector, t)
	require.NoError(t, err)

	// Set up contract mocks (same pattern as cron test)
	evmCfg := config.EVMs[0]

	balanceReaderMock := balance_reader.NewBalanceReaderMock(
		common.HexToAddress(evmCfg.BalanceReaderAddress),
		evmMock,
	)
	balanceReaderMock.GetNativeBalances = func(input balance_reader.GetNativeBalancesInput) ([]*big.Int, error) {
		// Return mock balance for each address (same number as input addresses)
		balances := make([]*big.Int, len(input.Addresses))
		for i := range input.Addresses {
			balances[i] = big.NewInt(750000000000000000) // 0.75 ETH in wei (higher for HTTP test)
		}
		return balances, nil
	}

	ierc20Mock := ierc20.NewIERC20Mock(
		common.HexToAddress(evmCfg.TokenAddress),
		evmMock,
	)
	ierc20Mock.TotalSupply = func() (*big.Int, error) {
		return big.NewInt(2000000000000000000), nil // 2 tokens with 18 decimals (higher for HTTP test)
	}

	evmMock.WriteReport = func(ctx context.Context, input *evm.WriteReportRequest) (*evm.WriteReportReply, error) {
		return &evm.WriteReportReply{
			TxHash: common.HexToHash("0x9876543210fedcba9876543210fedcba9876543210fedcba9876543210fedcba").Bytes(),
		}, nil
	}

	// Test with empty payload (should default to now)
	result, err := onHTTPTrigger(config, runtime, &http.Payload{Input: []byte{}})
	require.NoError(t, err)
	require.Equal(t, "2000000", result)

	// Test with JSON payload
	httpPayload := HTTPTriggerPayload{
		ExecutionTime: anyExecutionTime,
	}
	payloadBytes, err := json.Marshal(httpPayload)
	require.NoError(t, err)

	result2, err := onHTTPTrigger(config, runtime, &http.Payload{Input: payloadBytes})
	require.NoError(t, err)
	require.Equal(t, "2000000", result2)
}

//go:embed config.json
var configJson []byte

func makeTestConfig(t *testing.T) *Config {
	config := &Config{}
	require.NoError(t, json.Unmarshal(configJson, config))
	return config
}

func assertLogContains(t *testing.T, logs [][]byte, substr string) {
	for _, line := range logs {
		if strings.Contains(string(line), substr) {
			return
		}
	}
	t.Fatalf("Expected logs to contain substring %q, but it was not found in logs:\n%s",
		substr, strings.Join(func() []string {
			var logStrings []string
			for _, log := range logs {
				logStrings = append(logStrings, string(log))
			}
			return logStrings
		}(), "\n"))
}

File: starter-templates/custom-data-feed/cre-custom-data-feed-ts/my-workflow/workflow.test.ts
import { describe, expect } from 'bun:test'
import { cre, getNetwork, TxStatus } from '@chainlink/cre-sdk'
import { EvmMock, newTestRuntime, test } from '@chainlink/cre-sdk/test'
import type { Address } from 'viem'
import { BalanceReader } from '../contracts/evm/ts/generated/BalanceReader'
import { newBalanceReaderMock } from '../contracts/evm/ts/generated/BalanceReader_mock'
import { MessageEmitter } from '../contracts/evm/ts/generated/MessageEmitter'
import { newMessageEmitterMock } from '../contracts/evm/ts/generated/MessageEmitter_mock'
import { ReserveManager } from '../contracts/evm/ts/generated/ReserveManager'
import { newReserveManagerMock } from '../contracts/evm/ts/generated/ReserveManager_mock'
import { initWorkflow, onCronTrigger, onLogTrigger } from './workflow'

const CHAIN_SELECTOR = 16015286601757825753n // ethereum-testnet-sepolia
const BALANCE_READER = '0x7109709ECfa91a80626fF3989D68f67F5b1DD12D' as Address
const MESSAGE_EMITTER = '0x694AA1769357215DE4FAC081bf1f309aDC325306' as Address
const RESERVE_MANAGER = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48' as Address
const TOKEN_HOLDER = '0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984' as Address

describe('onCronTrigger', () => {
	test('reads native token balances on-chain via BalanceReader', async () => {
		const evmMock = EvmMock.testInstance(CHAIN_SELECTOR)
		const balanceReaderMock = newBalanceReaderMock(BALANCE_READER, evmMock)
		balanceReaderMock.getNativeBalances = () => [1000000000000000000n]

		const runtime = newTestRuntime()
		const network = getNetwork({
			chainFamily: 'evm',
			chainSelectorName: 'ethereum-testnet-sepolia',
			isTestnet: true,
		})
		expect(network).toBeDefined()
		if (!network) return

		const evmClient = new cre.capabilities.EVMClient(network.chainSelector.selector)
		const reader = new BalanceReader(evmClient, BALANCE_READER)
		const balances = reader.getNativeBalances(runtime, [TOKEN_HOLDER])
		expect(balances).toEqual([1000000000000000000n])
	})

	test('writes updated reserve data on-chain via ReserveManager', async () => {
		const evmMock = EvmMock.testInstance(CHAIN_SELECTOR)
		const reserveManagerMock = newReserveManagerMock(RESERVE_MANAGER, evmMock)
		reserveManagerMock.writeReport = () => ({
			txStatus: TxStatus.SUCCESS,
			txHash: new Uint8Array(32),
		})

		const runtime = newTestRuntime()
		const network = getNetwork({
			chainFamily: 'evm',
			chainSelectorName: 'ethereum-testnet-sepolia',
			isTestnet: true,
		})
		expect(network).toBeDefined()
		if (!network) return

		const evmClient = new cre.capabilities.EVMClient(network.chainSelector.selector)
		const manager = new ReserveManager(evmClient, RESERVE_MANAGER)
		const resp = manager.writeReportFromUpdateReserves(
			runtime,
			{ totalMinted: 1000n, totalReserve: 2000n },
			{ gasLimit: '500000' },
		)
		expect(resp.txStatus).toBe(TxStatus.SUCCESS)
	})

	test('throws when scheduledExecutionTime is missing', () => {
		const runtime = newTestRuntime()
		expect(() => onCronTrigger(runtime as any, {} as any)).toThrow(
			'Scheduled execution time is required',
		)
	})
})

describe('onLogTrigger', () => {
	test('returns decoded message from adapted trigger payload', () => {
		const runtime = newTestRuntime() as any

		const payload = {
			topics: [],
			data: {
				emitter: TOKEN_HOLDER as `0x${string}`,
				timestamp: 1709510400n,
				message: 'hello-from-adapted-trigger',
			},
		}

		const result = onLogTrigger(runtime, payload as any)
		expect(result).toBe('hello-from-adapted-trigger')
	})

	test('reads last emitted message from MessageEmitter contract via mock', async () => {
		const evmMock = EvmMock.testInstance(CHAIN_SELECTOR)
		const msgEmitterMock = newMessageEmitterMock(MESSAGE_EMITTER, evmMock)
		msgEmitterMock.getLastMessage = () => 'hello-from-mock'

		const runtime = newTestRuntime()
		const network = getNetwork({
			chainFamily: 'evm',
			chainSelectorName: 'ethereum-testnet-sepolia',
			isTestnet: true,
		})
		expect(network).toBeDefined()
		if (!network) return

		const evmClient = new cre.capabilities.EVMClient(network.chainSelector.selector)
		const emitter = new MessageEmitter(evmClient, MESSAGE_EMITTER)
		const msg = emitter.getLastMessage(runtime, TOKEN_HOLDER)
		expect(msg).toBe('hello-from-mock')
	})
})

describe('initWorkflow', () => {
	test('subscribes onCronTrigger to cron schedule and onLogTrigger to log filter', () => {
		const config = {
			schedule: '0 */10 * * * *',
			url: 'https://api.example.com/por',
			evms: [
				{
					tokenAddress: BALANCE_READER,
					porAddress: RESERVE_MANAGER,
					proxyAddress: RESERVE_MANAGER,
					balanceReaderAddress: BALANCE_READER,
					messageEmitterAddress: MESSAGE_EMITTER,
					chainSelectorName: 'ethereum-testnet-sepolia',
					gasLimit: '500000',
				},
			],
		}
		const handlers = initWorkflow(config)

		expect(handlers).toHaveLength(2)

		expect(handlers[0].fn).toBe(onCronTrigger)
		const cronTrigger = handlers[0].trigger as { config?: { schedule?: string } }
		expect(cronTrigger.config?.schedule).toBe(config.schedule)

		expect(handlers[1].fn).toBe(onLogTrigger)
		const logTrigger = handlers[1].trigger as {
			adapt: (raw: any) => any
			configAsAny: () => any
		}
		expect(typeof logTrigger.adapt).toBe('function')
		expect(typeof logTrigger.configAsAny).toBe('function')
	})
})
